File: daily/2025/368_Largest_Divisible_Subset.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> largestDivisibleSubset(vector<int>& nums) {
        int n = nums.size();
        sort(nums.begin(), nums.end());
        vector<int> dp(n, 1);
        vector<int> prev(n, -1);
        for (int i = 1; i < n; i++){
            for (int j = 0; j < i; j++){
                if (nums[i] % nums[j] == 0 && dp[j] + 1 > dp[i]){
                    dp[i] = dp[j] + 1;
                    prev[i] = j;
                }
            }
        }
        int max_len = 1;
        int max_index = 0;
        for (int i = 0; i < n; i++){
            if (dp[i] > max_len) max_len = dp[i], max_index = i;
        }
        vector<int> ans;
        while(max_index != -1){
            ans.push_back(nums[max_index]);
            max_index = prev[max_index];
        }
        return ans;
    }
};

int main(){
    vector<int> nums = {1,2,3};
    // vector<int> nums = {1,2,4,8};
    Solution S;
    vector<int> ans = S.largestDivisibleSubset(nums);
    for (int i : ans) cout << i << " ";
    cout << endl;
    return 0;
}

File: daily_python/368_Largest_Divisible_Subset.py
import os
import sys
from typing import List

class Solution:
    def largestDivisibleSubset(self, nums: List[int]) -> List[int]:
        nums = sorted(nums)
        n = len(nums)
        dp = [1] * n
        prev = [-1] * n
        largest_index = 0
        for i in range(n):
            for j in range(i):
                if nums[i] % nums[j] == 0 and dp[j] + 1 > dp[i]:
                    dp[i] = dp[j] + 1
                    prev[i] = j
            if dp[i] > dp[largest_index]:
                largest_index = i
        ans = []
        while largest_index != -1:
            ans.append(nums[largest_index])
            largest_index = prev[largest_index]
        return ans
    
def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            nums = list(map(int, lines[0].split('=')[1].strip()[1:-2].split(',')))
            dataset.append(nums)
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for nums in dataset:
        results.append(solution.largestDivisibleSubset(nums))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')